File: src/utils/paginate.ts
import { Pagination } from '../api/types/pagination.js';

export interface PageConfig {
  pageSize?: number;
  pageNumber?: number;
  limit?: number;
}

interface PaginateArgs<T> {
  pageConfig: PageConfig;
  getDataFn: (pagination: PageConfig) => Promise<{ pagination: Pagination; data: Array<T> }>;
}

export async function paginate<T>({ pageConfig, getDataFn }: PaginateArgs<T>): Promise<Array<T>> {
  const { pageSize, limit } = pageConfig;
  const { pagination, data } = await getDataFn(pageConfig);
  const result = [...data];

  let { totalAvailable, pageNumber } = pagination;
  while (totalAvailable > result.length && (!limit || limit > result.length)) {
    const { pagination: nextPagination, data: nextData } = await getDataFn({
      pageSize,
      pageNumber: pageNumber + 1,
      limit,
    });

    if (nextData.length === 0) {
      break;
    }

    ({ totalAvailable, pageNumber } = nextPagination);
    result.push(...nextData);
  }

  if (limit && limit < result.length) {
    result.length = limit;
  }

  return result;
}

File: src/tableau_cli/utils/paginate.py
from __future__ import annotations

from typing import Any, Callable, Optional


def paginate(
    *,
    page_size: Optional[int] = None,
    limit: Optional[int] = None,
    get_data_fn: Callable[[Optional[int], Optional[int]], dict[str, Any]],
) -> list[Any]:
    """Paginate through Tableau REST API results.

    get_data_fn(page_size, page_number) should return {"pagination": {...}, "data": [...]}.
    """
    response = get_data_fn(page_size, None)
    pagination = response["pagination"]
    result = list(response["data"])

    total_available: int = pagination["totalAvailable"]
    page_number: int = pagination["pageNumber"]

    while total_available > len(result) and (not limit or limit > len(result)):
        response = get_data_fn(page_size, page_number + 1)
        next_data = response["data"]

        if len(next_data) == 0:
            raise RuntimeError(
                f"No more data available. Last fetched page number: {page_number}, "
                f"Total available: {total_available}, Total fetched: {len(result)}"
            )

        pagination = response["pagination"]
        total_available = pagination["totalAvailable"]
        page_number = pagination["pageNumber"]
        result.extend(next_data)

    if limit and limit < len(result):
        result = result[:limit]

    return result
